packages/block-editor/src/utils/dom.js

```diff
@@ -57,3 +57,110 @@ export function getBlockClientId( node ) {
 
 	return blockNode.id.slice( 'block-'.length );
 }
+
+/**
+ * Calculates the union of two rectangles, and optionally constrains this union within a containerRect's
+ * left and right values.
+ * The function returns a new DOMRect object representing this union.
+ *
+ * @param {DOMRect}          rect1         First rectangle.
+ * @param {DOMRect}          rect2         Second rectangle.
+ * @param {DOMRectReadOnly?} containerRect An optional container rectangle. The union will be clipped to this rectangle.
+ * @return {DOMRect} Union of the two rectangles.
+ */
+export function rectUnion( rect1, rect2, containerRect ) {
+	let left = Math.min( rect1.left, rect2.left );
+	let right = Math.max( rect1.right, rect2.right );
+	const bottom = Math.max( rect1.bottom, rect2.bottom );
+	const top = Math.min( rect1.top, rect2.top );
+
+	/*
+	 * To calculate visible bounds using rectUnion, take into account the outer
+	 * horizontal limits of the container in which an element is supposed to be "visible".
+	 * For example, if an element is positioned -10px to the left of the window x value (0),
+	 * this function discounts the negative overhang because it's not visible and
+	 * therefore not to be counted in the visibility calculations.
+	 * Top and bottom values are not accounted for to accommodate vertical scroll.
+	 */
+	if ( containerRect ) {
+		left = Math.max( left, containerRect.left );
+		right = Math.min( right, containerRect.right );
+	}
+
+	return new window.DOMRect( left, top, right - left, bottom - top );
+}
```

Comment:
The approach makes sense but I don't really think this belongs in `rectUnion`. In a previous commit we had `rectIntersect` which we could bring back.

Reply:
https://github.com/WordPress/gutenberg/pull/64855